### PR TITLE
fix: Generate, but do not write to the spec, the default API server ELB name

### LINKED ISSUE #3003: V1.1.0 is not compatible with v1.0.0 release due to setting up an optional field in the controller
In our community meetings, we have decided to only do minor releases starting from v1 and do a major release for breaking changes. So, all v1.x releases need to be compatible with each other.

In v1.1.0, we introduced a bug by adding an optional field

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/3e61428d885f2eaeab1a7dffc4d17270a20c6675/api/v1beta1/awscluster_types.go#L156

and then setting it a value in the controller:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/3e61428d885f2eaeab1a7dffc4d17270a20c6675/pkg/cloud/services/elb/loadbalancer.go#L72

Now, external controllers that use CAPA API v1.0.0, is failing to update AWSCluster because admission webhook denies the request with `invalid: spec.controlPlaneLoadBalancer.name: Invalid value: "null": field is immutable` error

/kind bug
/priority critical-urgent
/milestone v1.2.0
/triage accepted

### PR DESCRIPTION
**Pull request by dlipovetsky**
**What type of PR is this?**
/kind  bug

**What this PR does / why we need it**:
Instead of writing the generated API server ELB name to the AWSCluster Spec, generates it at runtime. Writing the value to the spec inadvertently broke the ability of v1.0.0 controllers to update AWSCluster resource created by a v1.1.0 controller.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3003

**Special notes for your reviewer**:
An AWSCluster created by a v1.1.0 controller already has its generated name written to the Spec. As of this PR, the generated name will not be written to the Spec. I don't think this is an issue, but wanted to note it.

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
TODO
```

### CODE REVIEW

**Comment by sedefsavas** [comment]
v1alpha4 upgrade failed but may just be timed out, restarted the test in the[ other PR](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3005).
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/directory/pull-cluster-api-provider-aws-e2e/1468223140587900928

File: pkg/cloud/services/elb/loadbalancer_test.go [hunk 1]
```diff
@@ -41,6 +41,66 @@ import (
...
+					},
+				},
+				AWSCluster: &tt.awsCluster,
+			})
+			if err != nil {
+				t.Fatalf("failed to create scope: %s", err)
+			}
+
+			elbName, err := ELBName(scope)
+			if elbName != tt.expected {
```
[the remark this answers is not in the document]
Comment (dlipovetsky) [reply pkg/cloud/services/elb/loadbalancer_test.go:97]: Thanks, fixing!

**Comment by dlipovetsky** [comment]
I'd like to squash the fixups, just let me know when.

**Comment by sedefsavas** [comment]
There is a known issue with GPU e2e test, I am only observing v1alpha3/v1alpha4 --> v1beta1 upgrade.

v1alpha3 --> v1beta1 upgrade failed, waiting for the logs.

**Comment by sedefsavas** [comment]
/test pull-cluster-api-provider-aws-e2e

The error was:
`[4]     Error: failed to get provider components for the "kubeadm:v0.3.23" provider: failed to parse yaml: failed to unmarshal the 1st yaml document: "\ufeff<?xml version=\"1.0\" encoding=\"utf-8\"?><Error><Code>ServerBusy</Code><Message>Egress is over the account limit.\nRequestId:e2cccbac-701e-005c-0211-ec773d000000\nTime:2021-12-08T08:54:32.7332372Z</Message></Error>\n": error unmarshaling JSON: while decoding JSON: json: cannot unmarshal string into Go value of type map[string]interface {}`

**Comment by sedefsavas** [comment]
I am suspecting that the failures are because of AWS or networking issues. Because the tests passed in the other PR with the same code (GPU failure is a known issue):
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/3005/pull-cluster-api-provider-aws-e2e/1468329832155713536

**Comment by sedefsavas** [comment]
Upgrade tests passed, the one failed timed out.
@dlipovetsky you are good to squash.

**Comment by dlipovetsky** [comment]
The e2e passed: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/3004/pull-cluster-api-provider-aws-e2e/1468615298205618176. Squashing.

**Comment by dlipovetsky** [comment]
(I squashed the fixups, but I'd like to keep the 3 commits separate, for clarity)

**Comment by sedefsavas** [comment]
Thanks @dlipovetsky for the quick fix!!

**Pull request merged**